gossipsub.go

```diff
@@ -2033,6 +2048,44 @@ func (gs *GossipSubRouter) enqueueGossip(p peer.ID, ihave *pb.ControlIHave) {
 	gs.gossip[p] = gossip
 }
 
+func (gs *GossipSubRouter) announceMessage(topic string, msg *Message, expiry time.Time) {
+	// Get all peers in topic
+	tmap, ok := gs.p.topics[topic]
+	if !ok {
+		return
+	}
+
+	// Store message for IWANT retrieval in the message cache
+	msgID := gs.p.idGen.ID(msg)
+
+	// Send IHAVE to all topic peers (excluding direct peers, applying score threshold)
+	// Match the filtering logic from emitGossip
+	var gossipQueued bool
+	for p := range tmap {
+		if !gs.feature(GossipSubFeatureMesh, gs.peers[p]) {
+			continue
+		}
+		if gs.score.Score(p) < gs.gossipThreshold {
+			continue
+		}
+		gs.enqueueGossip(p, &pb.ControlIHave{
+			TopicID:    &topic,
+			MessageIDs: []string{msgID},
+		})
+		gossipQueued = true
+	}
+
+	if !gossipQueued {
+		return
+	}
+
+	// Track announcement in message cache for IWANT retrieval
+	gs.mcache.TrackAnn(msg, expiry)
```

Comment:
I wonder if we can just remove the msgID from the MessageCache? We seem to be calculating the message ID over and over again, which adds up unnecessary compute time. We can calculate once and memoize.

More importantly the msg ID function can be overriden per topic, so having a global msgID in the MessageCache is definitely a design flaw.

Reply:
WDYT @sukunrt?